### ISSUE TITLE
Nats stream backup does not create consumer backup when consumer replica < stream replica

### ISSUE THREAD

**Issue by sourabhaggrawal** [nats-io/natscli#631]
Problem
When creating backup of an R3 stream it does not create  backup of R1 consumer. 
But same works when consumer replica is as same as stream replica. 

```
% nats pub backup --count=100 "Hello"

100 / 100 [==============================================================================================================================================================================================================================================]    0s

 % nats c report
? Select a Stream test-backup
╭────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                Consumer report for test-backup with 1 consumers                                │
├─────────────────┬──────┬────────────┬──────────┬─────────────┬─────────────┬─────────────┬───────────┬─────────┤
│ Consumer        │ Mode │ Ack Policy │ Ack Wait │ Ack Pending │ Redelivered │ Unprocessed │ Ack Floor │ Cluster │
├─────────────────┼──────┼────────────┼──────────┼─────────────┼─────────────┼─────────────┼───────────┼─────────┤
│ backup-consumer │ Pull │ Explicit   │ 30.00s   │ 0           │ 0           │ 100 / 100%  │ 0         │ n1*     │
╰─────────────────┴──────┴────────────┴──────────┴─────────────┴─────────────┴─────────────┴───────────┴─────────╯


 % nats str backup --consumers test-backup backup1
Starting backup of Stream "test-backup" with 4.0 KiB

3.1 KiB/s [==========================================================================================================================================================================================>---------------------------------------------------]  78%

Received 3.1 KiB compressed data in 2 chunks for stream "test-backup" in 2ms, 15 KiB uncompressed
 % nats str delete
? Select a Stream test-backup
? Really delete Stream test-backup Yes
sourabhagrawal@C02CX20FML85 scripts % nats str restore backup1
Starting restore of Stream "test-backup" from file "backup1"

3.1 KiB/s [==============================================================================================================================================================================================================================================] 100%

Restored stream "test-backup" in 0s

Information for Stream  created 2022-11-14 19:55:39

             Subjects: backup
             Replicas: 3
              Storage: File

Options:

            Retention: Limits
     Acknowledgements: true
       Discard Policy: Old
     Duplicate Window: 2m0s
    Allows Msg Delete: true
         Allows Purge: true
       Allows Rollups: false

Limits:

     Maximum Messages: unlimited
  Maximum Per Subject: unlimited
        Maximum Bytes: unlimited
          Maximum Age: unlimited
 Maximum Message Size: unlimited
    Maximum Consumers: unlimited


Cluster Information:

                 Name: local
               Leader: n0
              Replica: n1, current, seen 0.00s ago
              Replica: n2, current, seen 0.00s ago

State:

             Messages: 100
                Bytes: 4.0 KiB
             FirstSeq: 1 @ 2022-11-14T14:27:36 UTC
              LastSeq: 100 @ 2022-11-14T14:27:36 UTC
     Active Consumers: 0
   Number of Subjects: 1
 % nats c report
? Select a Stream test-backup
╭─────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│                            Consumer report for test-backup with 0 consumers                             │
├──────────┬──────┬────────────┬──────────┬─────────────┬─────────────┬─────────────┬───────────┬─────────┤
│ Consumer │ Mode │ Ack Policy │ Ack Wait │ Ack Pending │ Redelivered │ Unprocessed │ Ack Floor │ Cluster │
├──────────┼──────┼────────────┼──────────┼─────────────┼─────────────┼─────────────┼───────────┼─────────┤
╰──────────┴──────┴────────────┴──────────┴─────────────┴─────────────┴─────────────┴───────────┴─────────╯
```

**Comment by ripienaar** [comment]
What version server are you on? 

```
$ nats s restore `pwd`
Starting restore of Stream "ORDERS_0" from file "/home/rip/backup/orders_0"

212 KiB/s [================================================] 100%

Restored stream "ORDERS_0" in 0s
```

```
$ nats c info ORDERS_0 R1
...
            Replicas: 1
...
```

**Comment by sourabhaggrawal** [comment]
Tried this on 2.9.6

**Comment by sourabhaggrawal** [comment]
> 

Sure let me know if I should create the issue in server

**Comment by sourabhaggrawal** [comment]
I cleaned up the existing, below is the new consumer I created and ran the backup and restore again. After restore this consumer did not show up.

Configuration:

        Durable Name: r1-file-con-backup
           Pull Mode: true
      Filter Subject: backup
      Deliver Policy: All
          Ack Policy: Explicit
            Ack Wait: 30s
       Replay Policy: Instant
     Max Ack Pending: 1,000
   Max Waiting Pulls: 512
            Replicas: 1

Cluster Information:

                Name: local
              Leader: n0

State:

   Last Delivered Message: Consumer sequence: 0 Stream sequence: 0
     Acknowledgment floor: Consumer sequence: 0 Stream sequence: 0
         Outstanding Acks: 0 out of maximum 1,000
     Redelivered Messages: 0
     Unprocessed Messages: 100
            Waiting Pulls: 0 of maximum 512

**Comment by ripienaar** [comment]
Download the utilities from here https://github.com/klauspost/compress/releases and do:

```
$ ./s2d stream.tar.s2
$ tar -tvf stream.tar
...
-rw------- nats/nats       247 2022-11-14 14:49 obs/R1/meta.inf
-rw------- nats/nats        16 2022-11-14 14:49 obs/R1/meta.sum
-rw------- nats/nats        11 2022-11-14 14:49 obs/R1/o.dat
```

Here I have a consumer `R1` in the backup, pls check if your consumer is there

**Comment by ripienaar** [comment]
Any more information on this one?

**Comment by sourabhaggrawal** [comment]
Hi @ripienaar, tried executing this and I am getting this error. 

> ./s2d backup3/stream.tar.s2
zsh: bad CPU type in executable: ./s2d

Downloaded this version -> https://github.com/klauspost/compress/releases/download/v1.15.12/s2-OSX_arm64_1.15.12.tar.gz

OS version - macOS 12.3.1

**Comment by sourabhaggrawal** [comment]
ok i was trying with wrong version and able to run with s2-OSX_x86_64_1.15.12.tar.gz

I tried the backup and restore again with 1 R3 file stream and 1 R1 file consumer of it, it worked as expected.
But when I create another R1 consumer but memory and then do the backup stream with consumer flag -> delete stream ->  restore stream, I don't see any consumer. Even file consumer was gone.
Used the tool as well you shared, below is the output of that.
-rw-------  0 nats   nats      411 Dec  5 10:55 meta.inf
-rw-------  0 nats   nats       16 Dec  5 10:55 meta.sum
-rw-------  0 nats   nats       17 Dec  5 10:55 msgs/1.idx
-rw-------  0 nats   nats        0 Dec  5 10:55 msgs/1.blk

**Comment by ripienaar** [comment]
Memory streams and consumers don’t support backups atm :(

**Comment by sourabhaggrawal** [comment]
My stream is file and consumer is memory, no problem if backup does not support memory consumer. But if cluster has combination of memory and file consumers, after backup and restore file based is also lost.

**Comment by ripienaar** [comment]
So weird, also cant reprodce that.  I make a Stream with 2 consumers - one mem and one file - and no problem.  

Though I think the problem here is the case where the leader of the R1 consumer is not also the stream leader, then the consumer backup will not be made as the stream leader doesnt have its data, this one I can reproduce.

**Comment by sourabhaggrawal** [comment]
Weird, I am able to reproduce it every time, although I did not check and focused on which one is stream or consumer leader. 
May be you can try creating multiple R1 consumers for any consumer to go on non stream leader node and then create a backup.

**Comment by ripienaar** [comment]
Yeah like I mentioned I was able to reproduce it with that scenario, its limited to that rather than a general case of all file consumers missing if a file one is present.  It's about the consumer not being known to the stream leader doing the backup.  So only when its a different R level than the stream and only when the consumer isnt on the leader.

I let the devs know

**Comment by ripienaar** [comment]
This is resolved on recent servers - maybe only 2.10, but i cant reproduce now on dev head
